### PR TITLE
Documentation of the security model

### PR DESCRIPTION
**Pull request by tgraf**
<a href='#crh-start'></a><a href='#crh-data-%7B%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
<a href='https://www.codereviewhub.com/cilium/cilium/pull/503?approve=1'></a>&nbsp;<a href='https://github.com/cilium/cilium/pull/503'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by aanm** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/cilium/cilium/pull/503'>Refresh</a>.<a href='#crh-update-none'></a>

**Review by aanm — COMMENTED** [review]
In the image, why does `role=` have a different color? Do you think the image should also contain the `source` concept?

File: Documentation/architecture.rst [hunk 1]
```diff
@@ -199,243 +235,792 @@ Example
...
+
+Every time a new pod with the label `role=frontend` or `role=backend` is either
+started or stopped, the rules on every cluster node which run any such pods
+must be updated by either adding or removing the corresponding IP address from
+the list of allowed IP addresses. In large distributed applications, this could
+imply updating thousands of cluster nodes multiple times per second depending
+on the churn rate of deployed pods. Worse, the starting of new `role=frontend`
+pods must be delayed until all servers running `role=backend` pods have been
+updated with the new security rules as otherwise connection attempts from the
+new pod could be mistakenly dropped. This makes scaling efficiently difficult.
```
Comment (borkmann) [inline Documentation/architecture.rst]: `This makes scaling efficiently difficult.` Perhaps `This makes it difficult to scale efficiently.`

File: Documentation/architecture.rst [hunk 2]
```diff
@@ -199,243 +235,792 @@ Example
...
+
+In order to avoid these complications which can limit scalability and
+flexibility, Cilium entirely separates security from network addressing.
+Instead, security is based on the identity of a pod. This identity can be
```
Comment (borkmann) [inline Documentation/architecture.rst]: Perhaps clear transition from labels to identity is a bit missing. Maybe `Instead, security is based on the identity of a pod, which is derived through labels.`

File: Documentation/architecture.rst [hunk 3]
```diff
@@ -199,243 +235,792 @@ Example
...
-part of the Cilium network.  This can be achieved by having a load-balancer container that both has a
-public IP on an externally reachable network and a private IP on a cilium network. However, many container
-orchestrater frameworks, like Kubernetes, have built in abstractions to handle this "ingress" load-balancing
-capability, which achieve the same effect that Cilium handles forwarding and security only for
-''internal'' traffic between different services.
+What is an Endpoint Identity?
+-----------------------------
+
+The identity of an endpoint is derived based on the labels associated with the
+pod or container. When a pod or container is started, Cilium will creats an
```
Comment (borkmann) [inline Documentation/architecture.rst]: `will creats`

File: Documentation/architecture.rst [hunk 4]
```diff
@@ -199,243 +235,792 @@ Example
...
+endpoint based on the event received by the container runtime to represent the
+pod or container on the network. As a next step, Cilium will resolve the
+identity of the endpoint created. Whenever the labels of the pod or container
+change, the identity is reconfirmed and automatically modified as required.
+
+Not all labels associated with a container or pod are meaningful when deriving
+the security identity. Labels may be used to store metadata such as the
+timestamp when a container was launched. Cilium requires to know which labels
+are meaningful and are subject to being be considered when deriving the
```
Comment (borkmann) [inline Documentation/architecture.rst]: `are subject to being be considered`

File: Documentation/architecture.rst [hunk 5]
```diff
@@ -199,243 +235,792 @@ Example
...
+identity. For this purpose, the user is required to specify a list of string
+prefixes of meaningful labels The standard behavior is to include all labels
```
Comment (borkmann) [inline Documentation/architecture.rst]: `labels The standard`

File: Documentation/architecture.rst [hunk 6]
```diff
@@ -199,243 +235,792 @@ Example
...
+which start with the prefix `id.`, e.g.  `id.service1`, `id.service2`,
+`id.groupA.service44`.
```
Comment (borkmann) [inline Documentation/architecture.rst]: Can it be overridden, if yes, with which command?

File: Documentation/architecture.rst [hunk 7]
```diff
@@ -199,243 +235,792 @@ Example
...
+identity of the transmitting endpoint is embedded into every network packet
+that is transmitted in between cluster nodes. The receiving cluster node can
+then extract the identity and verify whether a particular identity is allowed
+to communicate with any of the local endpoints.
+
+Default Security Policy
+-----------------------
+
+If no policy is loaded, the default policy is to allow all communication unless
+policy policy enforcement is explicitly enabled. As soon as the first policy
```
Comment (borkmann) [inline Documentation/architecture.rst]: `unless policy policy enforcement is`

File: Documentation/architecture.rst [hunk 8]
```diff
@@ -199,243 +235,792 @@ Example
...
+
+Allow Rules
+-----------
+
+This is the simplest rule type. The rule defines a list of labels which are
+allowed to consume whatever endpoints are covered by the coverage.
+
+If an endpoint transmits to another endpoint and the communication is not
+permitted by at least one *allow* rule, all packets of the connection will be
+silently dropped.
```
Comment (borkmann) [inline Documentation/architecture.rst]: Maybe cilium monitor could be mentioned here on how to debug drops and policy violations. But could be a future section for debugging/introspection perhaps. Could also have the links to the ascii cinema demos.

File: Documentation/architecture.rst [hunk 9]
```diff
@@ -199,243 +235,792 @@ Example
...
+| Field         | Type     | Description                                       |
++---------------+----------+---------------------------------------------------+
+| coverage      | Array of | List of labels which must match in order for this |
+|               | strings  | rule to be applied.                               |
++---------------+----------+---------------------------------------------------+
+| allow         | Array of | List of labels which are allowed to initiate a    |
+|               | strings  | connection to any endpoint covered by coverage.   |
+|               |          | The action is "accept" unless the label has the   |
+|               |          | prefix `!` in which case the action is "deny".    |
++---------------+----------+---------------------------------------------------+
```
Comment (borkmann) [inline Documentation/architecture.rst:655]: In the `readthedocs`, somehow this rightmost cell is cut off to fit the page, scrolling to the right by marking the text.

File: Documentation/architecture.rst [hunk 10]
```diff
@@ -199,243 +235,792 @@ Example
...
+incoming and outgoing requests. The enforcement point is defined by the
+location of the rules in either the "in-ports" or "out-ports" field of the
+Layer 4 policy rule.
+
+Unlike Layer 3 and Layer 4 policies, violation of Layer 7 rules does not result
+in packet drops. Instead, if possible, an access denied message such as an
+*HTTP 403 access denied* is sent back to the sending endpoint.
+
+TODO: describe rules
+
```
Comment (borkmann) [inline Documentation/architecture.rst:827]: Maybe link to ascii demo.

File: Documentation/architecture.rst [hunk 11]
```diff
@@ -199,243 +235,792 @@ Example
...
+Merging of Nodes
+^^^^^^^^^^^^^^^^
+
+TODO
+
+Policy Repository
+=================
+
+Policy rules imported into the Cilium agent are not shared with other compute
+node and are only enforced within the boundaries of the compute node. In order
```
Comment (borkmann) [inline Documentation/architecture.rst]: s/node/nodes/

**Review by borkmann — APPROVED** [review]
Looks great, couple of minor things commented.

File: Documentation/architecture.rst (hunk 10 above)
Comment (tgraf) [reply Documentation/architecture.rst:827]: Will do once it is linked

**Comment by tgraf** [comment]
> In the image, why does role= have a different color? Do you think the image should also contain the source concept?

The graphics can definitely be done better. Agreed that source needs to be documented as well. This is no way near complete. It's a start.

**Pull request merged**